core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/CompressRequestStage.java

```diff
@@ -56,23 +61,27 @@ public SdkHttpFullRequest.Builder execute(SdkHttpFullRequest.Builder input, Requ
 
         Compressor compressor = resolveCompressionType(context.executionAttributes());
 
-        // non-streaming
-        if (!isStreaming(context)) {
-            ContentStreamProvider wrappedProvider = input.contentStreamProvider();
-            ContentStreamProvider compressedStreamProvider = () -> compressor.compress(wrappedProvider.newStream());
-            input.contentStreamProvider(compressedStreamProvider);
-            input = updateContentEncodingHeader(input, compressor);
-            return updateContentLengthHeader(input);
+        // non-streaming OR transfer-encoding chunked
+        if (!isStreaming(context) || isTransferEncodingChunked(input)) {
+            compressEntirePayload(input, compressor);
+            updateContentEncodingHeader(input, compressor);
+            if (!isStreaming(context)) {
+                return updateContentLengthHeader(input);
+            } else {
+                return input.removeHeader("Content-Length");
```

Comment:
Why are we removing the "Content-Length" ?
Is this added anywhere else ? What if user has set some "Content-Length"

Reply:
We are only removing for streaming operations. We are sending in compressed chunks so we don't know the full length. Modeling tools throw an error for operations with streaming AND requiresLength